Blog post on SymPy Documentation (#794)
Quansight/Quansight-website: apps/labs/posts/sympy_documentation.md

   the CZI grant work. See below for more details. Splitting large pages
   hasn't been done yet due to technical difficulties with the Sphinx autodoc
   extension, as well as due to the fact that the large pages are now much
   easier to navigate with the new Sphinx theme (see below).

   The SymPy Live extension in the documentation was removed, as it was
   considered too much of a maintenance burden for the SymPy community to
   maintain. There is a new [SymPy Live](https://live.sympy.org/) shell that
   runs on [JupyterLite](https://jupyter.org/try-jupyter/lab/index.html)—that
   is, it runs entirely in the browser using
   [Pyodide](https://pyodide.org/en/stable/). We are hopeful that the
   JupyterLite community can come up with an equivalent SymPy Live-like
   extension that so that we can re-enable similar functionality in the SymPy
   documentation.

3. We were been able to identify some primary areas of documentation that
   to prioritize for writing new documentation guides.

Click if you wish to read the [full survey results](https://www.sympy.org/sympy-docs-survey/2021-docs-survey.html).

# Improved Sphinx Theme - Furo

> <!-- TODO: organize these pictures so that the text is a caption -->

Reviewer: These are actually parsed as MDX, so comments are spelled:

```suggestion
{/* TODO: organize these pictures so that the text is a caption */}
```

Author: We need to just address these comments before merging anyway. Do you know if there's a good way to do this? Otherwise, let's just leave them as images like they are now.

documentation. As you can see from the screenshot above from early 2021, the
SymPy documentation main page used to just consist of a long list of every
page in the documentation. The new Furo theme makes this list unnecessary, but
it also gave us an opportunity to explore how these pages could be explored
in a more logical way.

We decided to adopt the [Diátaxis](https://diataxis.fr/) framework for documentation
organization. Diátaxis splits documentation pages into one of
four categories, depending on whether the reader is interested in practical or
theoretical knowledge, and on whether they have study or work oriented goals.
As can be seen from the current documentation main page, the docs are now
organized into four categories: tutorials, how-to guides, explanations, and
API reference. We additionally added "installation" and "contributing" (see
below) as separate top-level categories. Installation is important enough to
warrant calling out documentation for it separately. Contribution documentation
is separate because it serves a separate audience, people who want to
contribute to SymPy, rather than people who are interested in using it.

> <!-- TODO: Add Diátaxis framework diagram here -->

Reviewer: Same, MDX comment:

```suggestion
{/* TODO: Add Diátaxis framework diagram here */}
```

Author: Looks like I can use the image from https://diataxis.fr/ is [CC-BY-SA](https://github.com/evildmp/diataxis-documentation-framework/blob/main/LICENSE.rst). Is it OK to use that? Otherwise, let's just not worry about including this diagram.

    def fdiff(self, argindex=1):
        return 1/self.args[0]
```

```py
>>> x = sympy.Symbol('x')
>>> log(1)
0
>>> log(x).diff(x)
1/x
```

SymPy now includes an extensive [how-to guide on defining custom symbolic
functions](https://docs.sympy.org/latest/guides/custom-functions.html). This
guide is useful to advanced users, but also this exact same
method is used to define the functions that are included in
SymPy itself. So this guide serves as both a guide to advanced end-users as
well as a guide to SymPy developers looking to define or extend one of the
functions that comes with SymPy.

## Guide on SymPy Best Practices

SymPy has many pitfalls, both for new users and advanced users. The new guide
on [best practices](https://docs.sympy.org/dev/explanation/best-practices.html) 
goes over some ways to avoid these pitfalls.

For example, one pitfall that many new SymPy users run into is using strings
as inputs to SymPy functions, like

```py
>>> from sympy import expand
>>> expand("(x**2 + x)/x")
x + 1
```

It's much better to define symbolic variables and create expressions directly,
like

```
>>> from sympy import symbols
>>>> x = symbols('x')
>>> expand((x**2 + x)/x)
x + 1
```

> The [best practices
> page](https://docs.sympy.org/dev/explanation/best-practices.html#avoid-string-inputs)
> outlines why this is a bad idea, as well as dozens of other best practices

Reviewer 2: If you wish to keep the link:

```suggestion
The best practices page outlines why one should [avoid string inputs](https://docs.sympy.org/dev/explanation/best-practices.html#avoid-string-inputs), 
as well as dozens of other best practices
```

description: 'SymPy's documentation has received many significant improvements over the past two years thanks to funding by the Chan Zuckerberg Initiative.'
category: []
---

The Chan Zuckerberg Initiative (CZI) has funded
[SymPy](https://www.sympy.org/) as part of Cycle 4 of its [Essential Open
Source Software for Science (EOSS)](https://chanzuckerberg.com/eoss/) grant
program. As part of this work, Aaron Meurer worked on improving the SymPy
documentation, with a focus on writing new narritave documentation guides.

SymPy is a mature project, and has over 1000 functions and classes. Most of
these functions and classes have API reference documentation in the form of
docstrings, but the SymPy documentation has historically been lacking in
long-form narrative documentation to supplement these reference docs.

In this post, I will go over some of the key documentation improvements that
were made over the course of the 2-year grant period. Note that the
documentation improvements were only one part of the CZI grant to improve
SymPy. Other SymPy developers were funded to improve the performance of SymPy,
and to improve its code generation capabilities. These improvements are

> discussed in other blog posts.

Author: The codegen blog post is here https://github.com/mechmotum/mechmotum.github.io/pull/138. I don't know when it will be published so we may need to just add the link to it later.

  The remaining were only intended for internal SymPy use and should not be used
  by user code.

  See https://docs.sympy.org/latest/explanation/active-deprecations.html#deprecated-sympy-core-compatibility
  for details.

  This has been deprecated since SymPy version 1.10. It
  will be removed in a future version of SymPy.
  ```

  These warning messages give detailed information on what is deprecated, what
  users can replace their code with, what version the deprecation was added
  in, and a link to a [documentation
  page](https://docs.sympy.org/latest/explanation/active-deprecations.html#deprecated-sympy-core-compatibility).

- All activate deprecations are listed in a [new page the
  documentation](https://docs.sympy.org/latest/explanation/active-deprecations.html).
  This page gives more details about each deprecation than would be appropriate
  to put in the deprecation message, including details on why each deprecation
  was made. The page also gives [helpful
  information](https://docs.sympy.org/dev/explanation/active-deprecations.html#silencing-sympy-deprecation-warnings)
  on how to silence deprecation warnings.

## Guide on Writing Custom Functions

> SymPy comes with hundreds of mathematical functions built-in. But it also

Author: My understanding (as I explained in Slack) is that "built-in" should be hyphenated as it is a single compound word. I'm not really sure what you're trying to say here about "added on". I'm talking about the functions that come with SymPy.

```

```py
>>> x = sympy.Symbol('x')
>>> log(1)
0
>>> log(x).diff(x)
1/x
```

SymPy now includes an extensive [how-to guide on defining custom symbolic
functions](https://docs.sympy.org/latest/guides/custom-functions.html). This
guide is useful to advanced users, but also this exact same
method is used to define the functions that are included in
SymPy itself. So this guide serves as both a guide to advanced end-users as
well as a guide to SymPy developers looking to define or extend one of the
functions that comes with SymPy.

## Guide on SymPy Best Practices

SymPy has many pitfalls, both for new users and advanced users. The new guide
on [best
practices](https://docs.sympy.org/dev/explanation/best-practices.html) goes
over some of the best practices that should be applied to avoid these
pitfalls.

For example, one pitfall that many new SymPy users run into is using strings
as inputs to SymPy functions, like

```py
>>> from sympy import expand
>>> expand("(x**2 + x)/x")
x + 1
```

It's much better to define symbolic variables and create expressions directly,
like

```
>>> from sympy import symbols
>>>> x = symbols('x')
>>> expand((x**2 + x)/x)
x + 1
```

The [best practices

> page](https://docs.sympy.org/dev/explanation/best-practices.html#avoid-string-inputs)

Author: This links to a specific section on the page.

> page](https://docs.sympy.org/dev/explanation/best-practices.html#avoid-string-inputs)

Reviewer 2: Yes, I noticed that. I would still remove it. I would even consider paring back more links from the post because I believe there are a few different places where you link to the same page, or different places on the same page. Links are great but each link also represents a fork in the reader's attention, away from your story, so I would consider them carefully (and/or perhaps create a footnotes section?). 

If you want to keep this link, then I have a different suggestion, which I will add in a separate comment.

Another way to think about this. If I were editing this blog post for print, I would look for ways to bring the point of the link into the article itself. So instead of linking to the specific best practice, I would maybe inline a screenshot or a block quote, excerpting that section right here into the article itself. Which you've kind of already done with the code sample above.
